### DISCUSSION TITLE
returning error in M when Lua code crashed.

### DISCUSSION THREAD

**Discussion by AlainDsc** [anet-be/mlua#12] [Q&A] [answered]
Dear,

At last my first discussion in this repository,
and - hopefully - not my last ;-)

In our M environment, we have some C plugins - you know - that accelerate some frequently used functions,
most of all string-oriented.
We are trying to replace them by Lua code, and the progress looks very promising.
( Lua code is sometimes even faster than C code)

Lua brings the opportunity to highly expand functionality in M. Just think about XML parsing, JSON, and many other stuff.

Actually it is quite fun : you can literally write software, and switch from M to the Lua - whenever the other language is better suited - .  For example : on our development system, we had more than 16.000.000 Lua calls from M in about 24 hours.

However, I have a question I didn't find the answer for:
Is there a possibility that M is aware of an error during Lua execution ?

In most cases, the reply variable just contains the error message. So, one has to be carefull, otherwise this message is used as the result of the function.

**Comment by berwynhoyt** [comment, marked as answer]
Hi Alain. The `mlua.lua()` return value tells you whether the result was an error. From the README:

> On error, `mlua.lua()` returns nonzero and the error message is sent to stdout or returned in .output (if >1 parameter supplied).


Does this do what you need?

**Reply by AlainDsc** [reply]
Hello Berwyn, but how can I reach this return value in M ?
I tried several things in M, like:

```
s ret=&mlua.lua("x = 3")
s ret=$$&mlua.lua("x = 3")
```
Nothings works. I probably miss something trivial.

**Reply by berwynhoyt** [reply]
Oh, I see your problem now. You need one $, not two: `s ret $&mlua.lua("x = 3")`

**Reply by AlainDsc** [reply]
Oh, gosh, thank you, my problem is solved.
I feel quite humble today.

**Comment by berwynhoyt** [comment]
Yes, your examples work for me. They both set ret=0 because "x = 3" is valid code.

The following also works for me, and demonstrates both a non-error and and error result:
```
YDB> write $&mlua.lua("x=3",.out)
0
YDB> write $&mlua.lua("x=<>",.out)
-1
YDB> write out
Lua: [string "mlua(code)"]:1: unexpected symbol near '<'
```

FYI: my versions:
```
YDB> w $&mlua.version,!,$ZYRELEASE
302
YottaDB r1.39 Linux x86_64
```
But it also works in previous YDB versions.
